Examples/Io/Root/src/RootTrackParameterWriter.cpp

```diff
@@ -144,69 +189,140 @@ ActsExamples::ProcessCode ActsExamples::RootTrackParameterWriter::writeT(
 
   // Loop over the estimated track parameters
   for (std::size_t iparams = 0; iparams < trackParams.size(); ++iparams) {
+    const auto& params = trackParams[iparams];
     // The reference surface of the parameters, i.e. also the reference surface
     // of the first space point
-    const auto& surface = trackParams[iparams].referenceSurface();
-    // The estimated bound parameters vector
-    const auto params = trackParams[iparams].parameters();
-    m_loc0 = params[Acts::eBoundLoc0];
-    m_loc1 = params[Acts::eBoundLoc1];
-    m_phi = params[Acts::eBoundPhi];
-    m_theta = params[Acts::eBoundTheta];
-    m_qop = params[Acts::eBoundQOverP];
-    m_time = params[Acts::eBoundTime];
-    m_p = std::abs(1.0 / m_qop);
-    m_pt = m_p * std::sin(m_theta);
-    m_eta = std::atanh(std::cos(m_theta));
+    const auto& surface = params.referenceSurface();
+
+    m_volumeId = surface.geometryId().volume();
+    m_layerId = surface.geometryId().layer();
+    m_surfaceId = surface.geometryId().sensitive();
+
+    m_loc0 = params.parameters()[Acts::eBoundLoc0];
+    m_loc1 = params.parameters()[Acts::eBoundLoc1];
+    m_phi = params.parameters()[Acts::eBoundPhi];
+    m_theta = params.parameters()[Acts::eBoundTheta];
+    m_qop = params.parameters()[Acts::eBoundQOverP];
+    m_time = params.parameters()[Acts::eBoundTime];
+
+    auto getError = [&params](std::size_t idx) -> float {
+      if (!params.covariance().has_value()) {
+        return NaNfloat;
+      }
+      const auto& cov = *params.covariance();
+      if (cov(idx, idx) < 0) {
+        return NaNfloat;
+      }
+      return std::sqrt(cov(idx, idx));
+    };
```

Comment:
I remember, that the debugging of the Writer functions was quite annoying, especially when there were FPEs, because they are too long.

Could it make sense to make real functions out of the lambdas, since we are doing actual calculations in there? Then it might be easier to find out which call failed.

Reply:
hmm but usually `gdb` gives you line numbers for such things, right?

Reply:
Right I never encountered problems with debugging lambdas

Reply:
Hmm but how can you see function names in normal logs either?

Reply:
Last time, I had problems due to an FPE, I read the monitoring file, which ended with something like "there was an FPE in this function" It was quite cumbersome, to find, which of the 300 lines was the reason for it. It would have saved me a lot of time, if we had shorter functions.

Reply:
I would generally recommend to trace such FPEs in `gdb`. That directly gives you the line where the error occurs. Because even in small functions there could be multiple divisions e.g., and then one again have to bisect this by print statements...

Reply:
Generally a stacktrace should give you the line number even inside the lambda. Sometimes it is not reliable but it can be reproduced locally. Also the function is written in a FPE safe way, that is the reason for the lambda.

Reply:
I just tried to reconstruct the problem I faced with overly long functions, but the logs were actually useful. I am fine to keep it like that and I will refactor in case I fail again in the future.